### ISSUE TITLE
Extract the neuralnets?

### ISSUE THREAD

**Issue by oysteijo** [jacobhilton/backgammon#1]
Hi!

Supercool project you got. Why have I not seen this before?

I'm not a OCaml coder, so I would love it if you could extract the neural networks from your project. I have a similar project back in my yard, and it would be supercool to compare for of your findings with my findings.

If the neural networks can be easily extracted to, say Keras readable or NNEF or ONNX, can you also describe the network inputs and outputs?

**Comment by jacobhilton** [comment]
Thanks for your interest. I have uploaded the logs and TensorFlow checkpoints for the experiments to GCS: `gs://jacobhilton/backgammon/saves/`

The checkpoint currently being used in the live demo is `large.5000.ckpt`, which is from the "Increasing the size of the neural network" experiment, using the largest network (with 5 hidden layers of size 400, relu activation, and the expanded board representation) after 5,000 training games. You can download this using [`gsutil`](https://github.com/GoogleCloudPlatform/gsutil/):
```
gsutil cp gs://jacobhilton/backgammon/saves/large.5000.ckpt .
```

The checkpoints have been set up to be read by the OCaml code, but are compatible with other TensorFlow code. For example, here is some TensorFlow 1.x code for restoring the checkpoint and evaluating the model on the opening board position:

```
import tensorflow as tf


def network(
    *,
    hidden_layer_sizes=[400] * 5,
    activation="relu",
    representation="expanded",
    epsilon_init=0.1,
):
    input_size = {"original": 196, "modified": 198, "expanded": 326}[representation]
    output_size = 1
    dtype = tf.float32
    input_placeholder = tf.placeholder(shape=[None, input_size], dtype=dtype)
    layer_size_pairs = zip(
        [input_size] + hidden_layer_sizes, hidden_layer_sizes + [output_size]
    )
    activation_function = {"sigmoid": tf.sigmoid, "relu": tf.nn.relu}[activation]
    node = input_placeholder
    for i, (size_from, size_to) in enumerate(layer_size_pairs):
        bias_var = tf.get_variable(
            f"bias_{i}",
            shape=[1, size_to],
            dtype=dtype,
            initializer=tf.zeros_initializer,
        )
        connected_var = tf.get_variable(
            f"connected_{i}",
            shape=[size_from, size_to],
            dtype=dtype,
            initializer=tf.random_normal_initializer(stddev=0.1),
        )
        node = tf.matmul(node, connected_var) + bias_var
        if i != len(hidden_layer_sizes):
            node = activation_function(node)
    logits = node
    model = tf.sigmoid(logits)
    return input_placeholder, model


with tf.Session() as sess:
    input_placeholder, model = network()
    saver = tf.train.Saver()
    saver.restore(sess, "large.5000.ckpt")
    print(sess.run(model, {input_placeholder: [[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]]}))
```

The "expanded" input representation used here is defined as follows:
- for i from 1 to 3:
    - if the first player has >= i counters on the bar then 1 else 0
- if the first player has >= 4 counters on the bar then (number of counters they have on the bar - 3) / 2 else 0
- for i from 1 to 15:
    - if the first player has >= i counters borne off then 1 else 0
- for each point starting from the second player's home board and ending in the first player's home board:
    - if the first player has exactly 1 counter on that point then 1 else 0
    - for i from 2 to 5:
      - if the first player has >= i counters on that point then 1 else 0
    - if the first player has >= 6 counters on that point then (number of counters they have on that point - 5) / 2 else 0
- repeat everything with the two players switched and the order of the numbers reversed (hence reversing the entire representation corresponds to switching the two players)

The model outputs its estimate of the probability that the first player will win the game if they play next.

Hopefully I've got all of that correct, but no guarantees – it's been a while since I touched any of this.

**Comment by oysteijo** [comment]
I'm able to download your checkpoints. I'll see how I can extract the parameter matrices. I also do understand the "expanded" input encoding. I think I will be able to integrate you networks into my systems. Hang on, and I'll see what I can do.

**Comment by jacobhilton** [comment]
You can extract the parameter matrices using the above script by putting something like this at the end:
```
    parameters = sess.run({v.name: v for v in tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES)})
```

**Comment by oysteijo** [comment]
`from tensorflow.python.training import py_checkpoint_reader`
Does the trick even better. :-)

**Comment by oysteijo** [comment]
Are all activation functions ReLU? Sigmoid at output?

**Comment by jacobhilton** [comment]
For the large architecture yes, sigmoid at output, ReLU everywhere else. (For the small and medium architectures, all activations are sigmoid.)

**Comment by oysteijo** [comment]
Hi again, @jacobhilton 
I have now managed to extract the neural networks and incorporated these in my own tool. However, I think I get some incorrect values out. I suspect I have something wrong in the calculation from board position -> neural network inputs. Can you please provide a correct array of neural network inputs from the backgammon initial position? Please also state the output value from the same array. This will help me a lot in my bug tracking. I'm using large.5000.ckpt.

**Comment by jacobhilton** [comment]
The script above includes the input representation for the opening board position:

```
[0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 1., 1., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]
```

You should get the output `0.525667` for this board position for the `large.5000.ckpt` checkpoint.

**Comment by oysteijo** [comment]
```
>>> nn = nnet.NeuralNet("../../oystein_octopus/src/hilton_evaluator/large.5000.npz")
>>> nn.predict(x)
array([0.525667], dtype=float32)
```
Thanks! I then got the neural network correct. I still have a bug in my board->input calculation.

**Comment by oysteijo** [comment]
Yohoo! I got it!
```
oystein@lt-143044:~/oystein_octopus/src/hilton_evaluator$ ./evaltool 4HPwATDgc/ABMA
  Position ID: 4HPwATDgc/ABMA
 +13-14-15-16-17-18------19-20-21-22-23-24-+
 | X           O    |   | O              X |
 | X           O    |   | O              X |
 | X           O    |   | O                |
 | X                |   | O                |
 | X                |   | O                |
v|                  |BAR|                  |
 | O                |   | X                |
 | O                |   | X                |
 | O           X    |   | X                |
 | O           X    |   | X              O |
 | O           X    |   | X              O |
 +12-11-10--9--8--7-------6--5--4--3--2--1-+

[ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ]
Static evaluation:     Win    Win_g   Win_bg   Lose_g  Lose_bg   Money eq.
                   0.52567  0.00000  0.00000  0.00000  0.00000    0.05133
```
Since there's only one output, I fill in the gammon and backgammon values with zeros. Gammon and backgammons should be implemented in the future. :-)

Thanks a lot. Can I take your evaluation and create a bot that plays on FIBS?

**Comment by jacobhilton** [comment]
> Can I take your evaluation and create a bot that plays on FIBS?

Sure, as long as FIBS allows that.

**Comment by oysteijo** [comment]
Thanks. What should I name the bot?

**Comment by jacobhilton** [comment]
Maybe CamlBot?

**Comment by oysteijo** [comment]
@jacobhilton Hi Jacob!

I have now set up a player named CamlBot on FIBS. It plays with your large.5000.ckpt network and only that. All positions uses the neural network. There is no lookahead and it only accepts one point matches. I will let it play a couple of week, I am not logging the games and I'm actually not even logging the results. We can only watch the rating adjustments as it play some matches.

**Comment by jacobhilton** [comment]
Thank you, that's exciting! According to my experiments, it should have a rating around 250 points lower than GNU Backgammon.

**Comment by oysteijo** [comment]
Let's see. It is still low on experience. An internet friend of mine in Germany, Hans-Jürgen Schaefer (Dankon87@gmx.de), maintains a backgammon bot rating list. He indicates that its rating will be about 1818, which is about the same as you indicate. Both Hans-Jürgen and I, love to play against it.

https://mustrum.de/blot.html
